### PR TITLE
Fix to SPI DMA interrupt miss issue; adds fractional division for UART baud rate

### PR DESCRIPTION
**Pull request by hchiang**
### Pull Request Overview

This pull request fixes the missing RX interrupt that occurs when the SPI clock is divided below ~140. It moves the RX channel setup before the TX channel setup.  

It also adds fractional baud rate division for the UART, which allows it to have correct output even when the baud rate and clock speeds are close together, such as 115200 baud rate under a 1MHz clock.

### Testing Strategy

This pull request was tested by running the spi_buf testcase in userland when RC1M, RCSYS, and DFLL are used as the system clock.

The UART change has been shown to print the correct output to console at a baud rate of 115200  when any clock is being used (RC1M, RCFAST, OSC0, DFLL, PLL, RC80M).

### Documentation Updated

- [x] Kernel: Updated the relevant files in `/docs`, or no updates are required.
- [x] Userland: Added/updated the application README, if needed.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by alevy** [comment]
Did moving `self.enable` in the SPI's `read_write_bytes` not do the trick?

I can think of why swapping the order of the DMA channle setups would also help, but it seems more fickly to me (though maybe I'm not thinking through it properly).

**Comment by hchiang** [comment]
If you move `enable` to the bottom of `read_write_bytes` the SPI just outputs two bytes and then stops.

File: chips/sam4l/src/spi.rs [hunk 1]
```diff
@@ -464,6 +465,17 @@ impl SpiHw {
         // depending on the presence of the read/write below
         self.transfers_in_progress.set(0);
 
+        // Only setup the RX channel if we were passed a read_buffer inside
+        // of the option. `map()` checks this for us.
```
Comment (brghena) [inline chips/sam4l/src/spi.rs]: Please add to this comment, explaining why this block of code needs to go before the other one. That way no one mistakenly moves it again when refactoring the code.

**Comment by hchiang** [comment]
Ok, so I figured out then enable issue. The DMA `do_transfer` function that is called on read and write consists of `prepare_transfer` and `start_transfer`. If we replace the calls in `read_write_bytes` to `do_transfer` with `prepare_transfer` and then call `start_transfer` on dma_read/write after `self.enable`, then spi works properly by just moving `self.enable`. Should we go with this solution instead? It's a little more lengthy but more understandable.

**Comment by hchiang** [comment]
After further testing it looks like moving `enable` only works for some instances, whereas switching the order of the `read_buffer` and `write_buffer` setup works for all clock speeds/baud rates, so I stuck with that change.

File: chips/sam4l/src/spi.rs (hunk 1 above)
Comment (alevy) [reply chips/sam4l/src/spi.rs]: I think this is addressed not, right?

**Pull request merged**